File: rad/engine/02_catalog/store/retention.go
package store

import (
	"context"
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"time"

	kv "github.com/Southclaws/rad/rad/engine/01_kv"
	"github.com/Southclaws/rad/rad/engine/01_kv/keyenc"
	"github.com/Southclaws/rad/rad/engine/02_catalog/model"
	"github.com/Southclaws/rad/rad/engine/reject"
)

const retentionPinPrefix = "/rad/catalog/retention_pin/"

func retentionPinKey(id string) []byte { return []byte(retentionPinPrefix + id) }

// SaveRetentionPin publishes one durable retention edge. Repeating the exact
// pin is idempotent; changing a pin in place is forbidden so a stale release
// cannot accidentally release a different resource.
func SaveRetentionPin(ctx context.Context, view kv.KV, pin model.RetentionPin) error {
	if pin.CreatedAt.IsZero() {
		pin.CreatedAt = time.Now().UTC()
	}
	if err := validateRetentionPin(pin); err != nil {
		return err
	}
	raw, err := json.Marshal(pin)
	if err != nil {
		return err
	}
	key := retentionPinKey(pin.ID)
	existing, ok, err := view.Get(ctx, key)
	if err != nil {
		return err
	}
	if ok {
		current, err := decodeRetentionPin(pin.ID, existing)
		if err != nil {
			return err
		}
		if sameRetentionPin(current, pin) {
			return nil
		}
		return fmt.Errorf("catalog: retention pin %q already protects a different resource", pin.ID)
	}
	return view.Put(ctx, key, raw)
}

func DeleteRetentionPin(ctx context.Context, view kv.KV, id string) error {
	if id == "" {
		return fmt.Errorf("catalog: retention pin ID must not be empty")
	}
	return view.Delete(ctx, retentionPinKey(id))
}

func GetRetentionPin(ctx context.Context, view kv.KV, id string) (model.RetentionPin, bool, error) {
	raw, ok, err := view.Get(ctx, retentionPinKey(id))
	if err != nil || !ok {
		return model.RetentionPin{}, ok, err
	}
	pin, err := decodeRetentionPin(id, raw)
	if err != nil {
		return model.RetentionPin{}, false, err
	}
	return pin, true, nil
}

func ListRetentionPins(ctx context.Context, view kv.KV) ([]model.RetentionPin, error) {
	start := []byte(retentionPinPrefix)
	it, err := view.Scan(ctx, start, keyenc.PrefixEnd(start))
	if err != nil {
		return nil, err
	}
	defer it.Close()
	var pins []model.RetentionPin
	for it.Next() {
		id := strings.TrimPrefix(string(it.Key()), retentionPinPrefix)
		pin, err := decodeRetentionPin(id, it.Value())
		if err != nil {
			return nil, err
		}
		pins = append(pins, pin)
	}
	return pins, it.Err()
}

func decodeRetentionPin(id string, raw []byte) (model.RetentionPin, error) {
	pin, err := decodeDurableJSON[model.RetentionPin]("retention pin", id, raw)
	if err != nil {
		return model.RetentionPin{}, err
	}
	if pin.ID != id {
		return model.RetentionPin{}, reject.Fail(
			reject.ReasonCatalogCorrupt,
			"catalog: retention pin key %q contains ID %q",
			id,
			pin.ID,
		)
	}
	if err := validateRetentionPin(pin); err != nil {
		return model.RetentionPin{}, reject.Mark(reject.ReasonCatalogCorrupt, err)
	}
	return pin, nil
}

func RetentionHorizons(ctx context.Context, view kv.KV) (model.RetentionHorizons, error) {
	pins, err := ListRetentionPins(ctx, view)
	if err != nil {
		return model.RetentionHorizons{}, err
	}
	grouped := make(map[model.RetentionResource]uint64)
	for _, pin := range pins {
		grouped[pin.Resource]++
	}
	var horizons model.RetentionHorizons
	for resource, pinCount := range grouped {
		horizon := model.RetentionHorizon{Resource: resource, PinCount: pinCount}
		switch resource.Kind {
		case model.RetentionTableDefinition, model.RetentionWriteProtocolDefinition:
			horizons.CatalogDefinitions = append(horizons.CatalogDefinitions, horizon)
		case model.RetentionDataSnapshot:
			horizons.DataSnapshots = append(horizons.DataSnapshots, horizon)
		case model.RetentionTransitionDeltas:
			horizons.TransitionDeltas = append(horizons.TransitionDeltas, horizon)
		case model.RetentionPhysicalTable, model.RetentionPhysicalColumn, model.RetentionPhysicalIndex:
			horizons.PhysicalArtifacts = append(horizons.PhysicalArtifacts, horizon)
		case model.RetentionTransitionDiagnostics:
			horizons.TransitionDiagnostics = append(horizons.TransitionDiagnostics, horizon)
		}
	}
	sortRetentionHorizons(horizons.CatalogDefinitions)
	sortRetentionHorizons(horizons.DataSnapshots)
	sortRetentionHorizons(horizons.TransitionDeltas)
	sortRetentionHorizons(horizons.PhysicalArtifacts)
	sortRetentionHorizons(horizons.TransitionDiagnostics)
	return horizons, nil
}

func sortRetentionHorizons(horizons []model.RetentionHorizon) {
	sort.Slice(horizons, func(i, j int) bool {
		a, b := horizons[i].Resource, horizons[j].Resource
		switch {
		case a.Kind != b.Kind:
			return a.Kind < b.Kind
		case a.TableID != b.TableID:
			return a.TableID < b.TableID
		case a.TableSchemaID != b.TableSchemaID:
			return a.TableSchemaID < b.TableSchemaID
		case a.ColumnID != b.ColumnID:
			return a.ColumnID < b.ColumnID
		case a.IndexID != b.IndexID:
			return a.IndexID < b.IndexID
		case a.DefinitionGeneration != b.DefinitionGeneration:
			return a.DefinitionGeneration < b.DefinitionGeneration
		case a.WriteProtocolGeneration != b.WriteProtocolGeneration:
			return a.WriteProtocolGeneration < b.WriteProtocolGeneration
		case a.TransitionID != b.TransitionID:
			return a.TransitionID < b.TransitionID
		default:
			return a.DataPosition < b.DataPosition
		}
	})
}

// RetentionBlocker returns one durable pin that prevents the candidate's
// physical work. An opaque retained data snapshot conservatively blocks every
// reclamation class until the KV layer grows a position-aware retention proof.
func RetentionBlocker(
	ctx context.Context,
	view kv.KV,
	reclamation model.Reclamation,
) (model.RetentionPin, bool, error) {
	pins, err := ListRetentionPins(ctx, view)
	if err != nil {
		return model.RetentionPin{}, false, err
	}
	for _, pin := range pins {
		if pin.Resource.Kind == model.RetentionDataSnapshot || pinBlocksReclamation(pin.Resource, reclamation) {
			return pin, true, nil
		}
	}
	return model.RetentionPin{}, false, nil
}

func pinBlocksReclamation(resource model.RetentionResource, reclamation model.Reclamation) bool {
	switch reclamation.Kind {
	case model.ReclamationTable:
		return resource.Kind == model.RetentionPhysicalTable && resource.TableID == reclamation.TableID
	case model.ReclamationColumn:
		return resource.Kind == model.RetentionPhysicalColumn &&
			resource.TableID == reclamation.TableID && resource.ColumnID == reclamation.ColumnID
	case model.ReclamationIndex:
		return resource.Kind == model.RetentionPhysicalIndex &&
			resource.TableID == reclamation.TableID && resource.IndexID == reclamation.IndexID
	case model.ReclamationTableDefinition:
		return resource.Kind == model.RetentionTableDefinition &&
			resource.TableSchemaID == reclamation.TableSchemaID &&
			resource.DefinitionGeneration == reclamation.DefinitionGeneration
	case model.ReclamationWriteProtocolDefinition:
		return resource.Kind == model.RetentionWriteProtocolDefinition &&
			resource.TableID == reclamation.TableID &&
			resource.WriteProtocolGeneration == reclamation.WriteProtocolGeneration
	case model.ReclamationTransitionDeltas:
		return resource.Kind == model.RetentionTransitionDeltas &&
			resource.TransitionID == reclamation.TransitionID
	case model.ReclamationCancelledIndex, model.ReclamationFailedIndex:
		return resource.Kind == model.RetentionTransitionDeltas &&
			resource.TransitionID == reclamation.TransitionID ||
			resource.Kind == model.RetentionPhysicalIndex &&
				resource.TableID == reclamation.TableID && resource.IndexID == reclamation.IndexID
	case model.ReclamationReplacedColumn, model.ReclamationCancelledReplacement, model.ReclamationFailedReplacement:
		return resource.Kind == model.RetentionPhysicalColumn &&
			resource.TableID == reclamation.TableID && resource.ColumnID == reclamation.ColumnID
	case model.ReclamationConstraintValidation:
		return false
	default:
		return false
	}
}

func validateRetentionPin(pin model.RetentionPin) error {
	if pin.ID == "" || pin.OwnerKind == "" || pin.OwnerID == "" || pin.Resource.Kind == "" {
		return fmt.Errorf("catalog: invalid retention pin %+v", pin)
	}
	if pin.CreatedAt.IsZero() {
		return fmt.Errorf("catalog: retention pin %q has no creation timestamp", pin.ID)
	}
	switch pin.OwnerKind {
	case model.RetentionOwnerPreparedPlan, model.RetentionOwnerDataSnapshot, model.RetentionOwnerReplica,
		model.RetentionOwnerCDC, model.RetentionOwnerTransition, model.RetentionOwnerSchemaWorker,
		model.RetentionOwnerPhysicalReader:
	default:
		return fmt.Errorf("catalog: retention pin %q has unknown owner kind %q", pin.ID, pin.OwnerKind)
	}
	resource := pin.Resource
	switch resource.Kind {
	case model.RetentionTableDefinition:
		if resource.TableSchemaID == 0 || resource.DefinitionGeneration == 0 {
			return fmt.Errorf("catalog: retention pin %q has an incomplete table definition", pin.ID)
		}
	case model.RetentionWriteProtocolDefinition:
		if resource.TableID == "" || resource.WriteProtocolGeneration == 0 {
			return fmt.Errorf("catalog: retention pin %q has an incomplete write protocol", pin.ID)
		}
	case model.RetentionDataSnapshot:
		if resource.DataPosition == "" {
			return fmt.Errorf("catalog: retention pin %q has no data position", pin.ID)
		}
	case model.RetentionTransitionDeltas, model.RetentionTransitionDiagnostics:
		if resource.TransitionID == "" {
			return fmt.Errorf("catalog: retention pin %q has no transition identity", pin.ID)
		}
	case model.RetentionPhysicalTable:
		if resource.TableID == "" {
			return fmt.Errorf("catalog: retention pin %q has no physical table identity", pin.ID)
		}
	case model.RetentionPhysicalColumn:
		if resource.TableID == "" || resource.ColumnID == "" {
			return fmt.Errorf("catalog: retention pin %q has an incomplete physical column", pin.ID)
		}
	case model.RetentionPhysicalIndex:
		if resource.TableID == "" || resource.IndexID == "" {
			return fmt.Errorf("catalog: retention pin %q has an incomplete physical index", pin.ID)
		}
	default:
		return fmt.Errorf("catalog: retention pin %q has unknown resource kind %q", pin.ID, resource.Kind)
	}
	return nil
}

func sameRetentionPin(a, b model.RetentionPin) bool {
	return a.ID == b.ID && a.OwnerKind == b.OwnerKind && a.OwnerID == b.OwnerID && a.Resource == b.Resource
}

func TransitionDiagnosticsPinned(ctx context.Context, view kv.KV, transitionID string) (bool, error) {
	pins, err := ListRetentionPins(ctx, view)
	if err != nil {
		return false, err
	}
	for _, pin := range pins {
		if pin.Resource.Kind == model.RetentionTransitionDiagnostics &&
			pin.Resource.TransitionID == transitionID {
			return true, nil
		}
	}
	return false, nil
}

File: src/engine/02_catalog/store/retention.rs
use std::collections::HashMap;

use bytes::Bytes;

use crate::engine::catalog::identity::{RetentionPinId, TransitionId};
use crate::engine::catalog::model::{
    Reclamation, ReclamationKind, RetentionHorizon, RetentionHorizons, RetentionPin,
    RetentionResource, RetentionResourceKind, Timestamp,
};
use crate::engine::catalog::{Error, ErrorKind, Result};
use crate::engine::kv::KvView;

use super::durable_json::{decode, encode};
use super::{map_kv, prefix_range};

const RETENTION_PIN_PREFIX: &str = "/rad/catalog/retention_pin/";

fn retention_pin_key(id: &RetentionPinId) -> Vec<u8> {
    format!("{RETENTION_PIN_PREFIX}{id}").into_bytes()
}

pub async fn save_retention_pin<V: KvView + ?Sized>(
    view: &mut V,
    mut pin: RetentionPin,
    now: Timestamp,
) -> Result<()> {
    if pin.created_at.is_zero() {
        pin.created_at = now;
    }
    validate_retention_pin(&pin)?;
    let key = retention_pin_key(&pin.id);
    if let Some(raw) = view.get(&key).await.map_err(map_kv)? {
        let current = decode_retention_pin(pin.id.as_str(), &raw)?;
        if current.id == pin.id
            && current.owner_kind == pin.owner_kind
            && current.owner_id == pin.owner_id
            && current.resource == pin.resource
        {
            return Ok(());
        }
        return Err(Error::message(
            ErrorKind::Conflict,
            format!(
                "catalog: retention pin {:?} already protects a different resource",
                pin.id
            ),
        ));
    }
    let raw = encode("retention pin", pin.id.as_str(), &pin)?;
    view.put(Bytes::from(key), Bytes::from(raw))
        .await
        .map_err(map_kv)
}

pub async fn delete_retention_pin<V: KvView + ?Sized>(
    view: &mut V,
    id: &RetentionPinId,
) -> Result<()> {
    if id.is_empty() {
        return Err(Error::message(
            ErrorKind::InvalidInput,
            "catalog: retention pin ID must not be empty",
        ));
    }
    view.delete(&retention_pin_key(id)).await.map_err(map_kv)
}

pub async fn get_retention_pin<V: KvView + ?Sized>(
    view: &mut V,
    id: &RetentionPinId,
) -> Result<Option<RetentionPin>> {
    let Some(raw) = view.get(&retention_pin_key(id)).await.map_err(map_kv)? else {
        return Ok(None);
    };
    decode_retention_pin(id.as_str(), &raw).map(Some)
}

pub async fn list_retention_pins<V: KvView + ?Sized>(view: &mut V) -> Result<Vec<RetentionPin>> {
    let prefix = RETENTION_PIN_PREFIX.as_bytes();
    let mut iterator = view.scan(prefix_range(prefix)).await.map_err(map_kv)?;
    let mut pins = Vec::new();
    while let Some(entry) = iterator.next().await.map_err(map_kv)? {
        let id = entry
            .key
            .strip_prefix(prefix)
            .and_then(|value| std::str::from_utf8(value).ok())
            .ok_or_else(|| {
                Error::message(
                    ErrorKind::CatalogCorrupt,
                    "catalog: malformed retention key",
                )
            })?;
        pins.push(decode_retention_pin(id, &entry.value)?);
    }
    Ok(pins)
}

fn decode_retention_pin(id: &str, raw: &[u8]) -> Result<RetentionPin> {
    let pin: RetentionPin = decode("retention pin", id, raw)?;
    if pin.id.as_str() != id {
        return Err(Error::message(
            ErrorKind::CatalogCorrupt,
            format!("catalog: retention pin key {id:?} contains ID {:?}", pin.id),
        ));
    }
    validate_retention_pin(&pin)
        .map_err(|error| Error::message(ErrorKind::CatalogCorrupt, error.to_string()))?;
    Ok(pin)
}

pub async fn retention_horizons<V: KvView + ?Sized>(view: &mut V) -> Result<RetentionHorizons> {
    let mut grouped = HashMap::<RetentionResource, u64>::new();
    for pin in list_retention_pins(view).await? {
        *grouped.entry(pin.resource).or_default() += 1;
    }
    let mut horizons = RetentionHorizons::default();
    for (resource, pin_count) in grouped {
        let group = match resource.kind {
            RetentionResourceKind::TableDefinition
            | RetentionResourceKind::WriteProtocolDefinition => &mut horizons.catalog_definitions,
            RetentionResourceKind::DataSnapshot => &mut horizons.data_snapshots,
            RetentionResourceKind::TransitionDeltas => &mut horizons.transition_deltas,
            RetentionResourceKind::PhysicalTable
            | RetentionResourceKind::PhysicalColumn
            | RetentionResourceKind::PhysicalIndex => &mut horizons.physical_artifacts,
            RetentionResourceKind::TransitionDiagnostics => &mut horizons.transition_diagnostics,
        };
        group.push(RetentionHorizon {
            resource,
            pin_count,
        });
    }
    for group in [
        &mut horizons.catalog_definitions,
        &mut horizons.data_snapshots,
        &mut horizons.transition_deltas,
        &mut horizons.physical_artifacts,
        &mut horizons.transition_diagnostics,
    ] {
        group.sort_by(|left, right| left.resource.cmp(&right.resource));
    }
    Ok(horizons)
}

pub async fn retention_blocker<V: KvView + ?Sized>(
    view: &mut V,
    reclamation: &Reclamation,
) -> Result<Option<RetentionPin>> {
    Ok(list_retention_pins(view).await?.into_iter().find(|pin| {
        pin.resource.kind == RetentionResourceKind::DataSnapshot
            || pin_blocks_reclamation(&pin.resource, reclamation)
    }))
}

fn pin_blocks_reclamation(resource: &RetentionResource, reclamation: &Reclamation) -> bool {
    match reclamation.kind {
        ReclamationKind::Table => {
            resource.kind == RetentionResourceKind::PhysicalTable
                && resource.table_id == reclamation.table_id
        }
        ReclamationKind::Column
        | ReclamationKind::ReplacedColumn
        | ReclamationKind::CancelledReplacement
        | ReclamationKind::FailedReplacement => {
            resource.kind == RetentionResourceKind::PhysicalColumn
                && resource.table_id == reclamation.table_id
                && resource.column_id == reclamation.column_id
        }
        ReclamationKind::Index => {
            resource.kind == RetentionResourceKind::PhysicalIndex
                && resource.table_id == reclamation.table_id
                && resource.index_id == reclamation.index_id
        }
        ReclamationKind::TableDefinition => {
            resource.kind == RetentionResourceKind::TableDefinition
                && resource.table_schema_id == reclamation.table_schema_id
                && resource.definition_generation == reclamation.definition_generation
        }
        ReclamationKind::WriteProtocolDefinition => {
            resource.kind == RetentionResourceKind::WriteProtocolDefinition
                && resource.table_id == reclamation.table_id
                && resource.write_protocol_generation == reclamation.write_protocol_generation
        }
        ReclamationKind::TransitionDeltas => {
            resource.kind == RetentionResourceKind::TransitionDeltas
                && resource.transition_id == reclamation.transition_id
        }
        ReclamationKind::CancelledIndex | ReclamationKind::FailedIndex => {
            (resource.kind == RetentionResourceKind::TransitionDeltas
                && resource.transition_id == reclamation.transition_id)
                || (resource.kind == RetentionResourceKind::PhysicalIndex
                    && resource.table_id == reclamation.table_id
                    && resource.index_id == reclamation.index_id)
        }
        ReclamationKind::ConstraintValidation => false,
    }
}

fn validate_retention_pin(pin: &RetentionPin) -> Result<()> {
    if pin.id.is_empty() || pin.owner_id.is_empty() || pin.created_at.is_zero() {
        return Err(Error::message(
            ErrorKind::InvalidInput,
            format!("catalog: invalid retention pin {:?}", pin.id),
        ));
    }
    let resource = &pin.resource;
    let complete = match resource.kind {
        RetentionResourceKind::TableDefinition => {
            resource.table_schema_id.is_some() && !resource.definition_generation.is_zero()
        }
        RetentionResourceKind::WriteProtocolDefinition => {
            !resource.table_id.is_empty() && !resource.write_protocol_generation.is_zero()
        }
        RetentionResourceKind::DataSnapshot => !resource.data_position.is_empty(),
        RetentionResourceKind::TransitionDeltas | RetentionResourceKind::TransitionDiagnostics => {
            !resource.transition_id.is_empty()
        }
        RetentionResourceKind::PhysicalTable => !resource.table_id.is_empty(),
        RetentionResourceKind::PhysicalColumn => {
            !resource.table_id.is_empty() && !resource.column_id.is_empty()
        }
        RetentionResourceKind::PhysicalIndex => {
            !resource.table_id.is_empty() && !resource.index_id.is_empty()
        }
    };
    if !complete {
        return Err(Error::message(
            ErrorKind::InvalidInput,
            format!(
                "catalog: retention pin {:?} has an incomplete resource",
                pin.id
            ),
        ));
    }
    Ok(())
}

pub async fn transition_diagnostics_pinned<V: KvView + ?Sized>(
    view: &mut V,
    transition_id: &TransitionId,
) -> Result<bool> {
    Ok(list_retention_pins(view).await?.iter().any(|pin| {
        pin.resource.kind == RetentionResourceKind::TransitionDiagnostics
            && &pin.resource.transition_id == transition_id
    }))
}

#[cfg(test)]
mod tests {
    use crate::engine::catalog::identity::{CatalogVersion, DefinitionGeneration, SchemaId};
    use crate::engine::catalog::model::{
        ReclamationKind, RetentionOwnerKind, RetentionResourceKind,
    };
    use crate::engine::kv::slatedb;
    use crate::engine::kv::{Kv, TransactionalKv};

    use super::*;

    fn resource() -> RetentionResource {
        RetentionResource {
            kind: RetentionResourceKind::TableDefinition,
            table_id: Default::default(),
            table_schema_id: Some(SchemaId::new(1).unwrap()),
            column_id: Default::default(),
            index_id: Default::default(),
            definition_generation: DefinitionGeneration::from(2),
            write_protocol_generation: Default::default(),
            transition_id: Default::default(),
            data_position: Default::default(),
        }
    }

    #[tokio::test]
    async fn pins_are_idempotent_grouped_and_block_exact_reclamation() {
        let mut database = slatedb::Store::memory("catalog-retention").await.unwrap();
        for id in ["p1", "p2"] {
            save_retention_pin(
                &mut database,
                RetentionPin {
                    id: id.into(),
                    owner_kind: RetentionOwnerKind::PreparedPlan,
                    owner_id: id.into(),
                    resource: resource(),
                    created_at: Timestamp::default(),
                },
                Timestamp::test_value(),
            )
            .await
            .unwrap();
        }
        assert_eq!(
            retention_horizons(&mut database)
                .await
                .unwrap()
                .catalog_definitions[0]
                .pin_count,
            2
        );
        let mut reclamation = Reclamation::pending(
            "r1",
            ReclamationKind::TableDefinition,
            CatalogVersion::from(3),
            Timestamp::test_value(),
        );
        reclamation.table_schema_id = Some(SchemaId::new(1).unwrap());
        reclamation.definition_generation = DefinitionGeneration::from(2);
        assert!(
            retention_blocker(&mut database, &reclamation)
                .await
                .unwrap()
                .is_some()
        );

        let original = RetentionPinId::from("p1");
        let raw = Kv::get(&database, &retention_pin_key(&original))
            .await
            .unwrap()
            .unwrap();
        Kv::put(
            &database,
            Bytes::from(retention_pin_key(&RetentionPinId::from("alias"))),
            raw,
        )
        .await
        .unwrap();
        assert_eq!(
            list_retention_pins(&mut database).await.unwrap_err().kind(),
            ErrorKind::CatalogCorrupt
        );
        database.close().await.unwrap();
    }
}
